server/src/tracing.ts

```diff
@@ -0,0 +1,202 @@
+export type { Span } from "@sentry/tracing";
+
+interface SpanOptions extends SpanContext {
+  parentSpan?: Span;
+  timeout?: number;
+}
+
+// Monkey-patch Transaction to add an event to notify listeners (in our case,
+// child spans) when the transaction is finishing. This is super hacky.
+//
+// The most recent release of the Sentry SDK has a `finishTransaction` event
+// on the hub's client, but there are a lot of guard clauses anywhere the client
+// gets used internally, and I'm not sure how reliable it is for this use case.
+// We want to wrap up before the transaction finishes, not when a client that
+// may-or-may-not exist depending on configuring is finishing a transaction.
+// (Also, the transaction's `_hub` is a nullable private property, so it would
+// still be hacky to grab it and add a listener anyway.)
+interface PatchedTransaction extends Transaction {
+  _onFinish(listener: (...args: any[]) => void): void;
+  _emitter?: EventEmitter;
+}
+
+const _transactionFinish = Transaction.prototype.finish;
+Transaction.prototype.finish = function finish(
+  this: PatchedTransaction,
+  endTimestamp?: number
+) {
+  this._emitter?.emit("finish", this);
+  return _transactionFinish.call(this, endTimestamp);
+};
+
+// @ts-expect-error: _onFinish doesn't exist; we're adding it.
+Transaction.prototype._onFinish = function onFinish(
+  this: PatchedTransaction,
+  listener: (...args: any[]) => void
+) {
+  if (!this._emitter) this._emitter = new EventEmitter();
+  return this._emitter.once("finish", listener);
+};
+
+function getParentSpan(span: Span) {
+  if (span.parentSpanId) {
+    return span.spanRecorder?.spans?.find(
+      (s) => s.spanId === span.parentSpanId
+    );
+  }
+  return null;
+}
+
+/**
+ * Start a tracing span. The returned span should be explicitly ended with
+ * `finishSpan`. The created span will be a child of whatever span is currently
+ * active (and then become the current span itself), or if there is no current
+ * span or transaction, this will start one for you.
+ *
+ * Alternatively, can explicitly pass an actual span object to be the parent:
+ * `startSpan({ parentSpan: yourSpan })`. In this case, the new span won't
+ * automatically become the global "current" span.
+ *
+ * Spans created this way will be automatically canceled when their transaction
+ * finishes (Sentry will drop any unfinished spans). Alternatively, you can
+ * set `timeout` to a number of milliseconds, and the span will be canceled
+ * after that time (this should prevent you from accidentally leaving a span
+ * open forever). Canceled spans have a non-ok status set and a `cancel` tag
+ * with a reason.
+ *
+ * More on why spans need canceling:
+ * https://github.com/getsentry/sentry-javascript/issues/4165#issuecomment-971424754
+ */
+export function startSpan(options: SpanOptions): Span {
+  let { parentSpan, timeout, ...spanOptions } = options;
+
+  let scope;
+  if (!parentSpan) {
+    scope = Sentry.getCurrentHub().getScope();
```

Comment:
Given your comment above about `_hub` being nullable, are you sure `Sentry.getCurrentHub()` will always return an object that `getScope()` can be called on?

Reply:
Yeah, this part is kinda complicated: `getCurrentHub()` always returns a hub (if there’s not one, it creates one). It’s the *transaction’s* (or the parent span’s transaction’s) `_hub` property that may or may not be set, because the transaction is given its hub at some weird time in its lifecycle, and I think it also depends on sampling rules and other settings. (Also the hub may or may not have a client, also depending on configuration! One of Sentry’s SDK devs told me they are both reliable if the transaction has been sampled, but then you have to check *that,* and who knows what future internal refactoring in the Sentry SDK could change that rule.)

(Also, `getCurrentHub()` returns a different hub for every async domain [at some point they’re going to upgrade to async hooks/storage] so Sentry can have concurrent scopes, like for two requests happening at the same time. So it matters that we use the right scope, since the parent span’s transaction’s scope might not be from the current async domain, so choosing when to get a scope via which method is a whole thing. 😬)